### PR TITLE
Add comments to handler for async handshake

### PR DESCRIPTION
**Pull request by AdityaSripal**
Writeup in the comments of how to compose the utility functions provided by IBC to implement async handshakes

closes: #732

cc: @michaelfig

### CODE REVIEW

**Review by sangier — APPROVED** [review]
LGTM!

File: spec/core/ics-026-routing-module/README.md [hunk 1]
```diff
@@ -610,19 +652,30 @@ interface ChanCloseConfirm {
...
       datagram.channelIdentifier,
       datagram.proofInit,
       datagram.proofHeight
     )
+    // SYNCHRONOUS: the following calls happen syncrhonously with the call above
+    // ASYNCHRONOUS: the module callback will be called at a time later than the channel handler
+    // in this case, the channel identifier will be stored with a sentinel value so it is not taken
+    // by a new channel handshake and the capability is reserved for the application module.
+    // When the module eventually executes its callback it must call writeChannel so that the channel
+    // can be written into an OPEN state and the handshake can proceed on the counterparty.
```
Comment (michaelfig) [inline spec/core/ics-026-routing-module/README.md:667]:
```suggestion
    // can be written into a CLOSED state and the handshake can proceed on the counterparty.
```

File: spec/core/ics-026-routing-module/README.md [hunk 2]
```diff
@@ -447,6 +447,12 @@ function handleChanOpenInit(datagram: ChanOpenInit) {
     datagram.portIdentifier,
     datagram.counterpartyPortIdentifier
   )
+  // SYNCHRONOUS: the following calls happen syncrhonously with the call above
```
Comment (michaelfig) [inline spec/core/ics-026-routing-module/README.md:450]: Minor typo here and in the below `SYNCHRONOUS:` comments.
```suggestion
  // SYNCHRONOUS: the following calls happen synchronously with the call above
```

**Review by michaelfig — APPROVED** [review]
A few minor copy-pasta typos.

**Comment by michaelfig** [comment]
@AdityaSripal, for async version negotiation in `ibc-go`, will I have to patch `core/keeper/msg_server.go`'s `ChannelOpenTry` with the following?

```go
	// Perform application logic callback
	version, err := cbs.OnChanOpenTry(ctx, msg.Channel.Ordering, msg.Channel.ConnectionHops, msg.PortId, channelID, cap, msg.Channel.Counterparty, msg.CounterpartyVersion)
	if err != nil {
		return nil, sdkerrors.Wrap(err, "channel open try callback failed")
	}

	// FIGME: If version is the empty string, expect the app to call WriteOpenTryChannel explicitly, possibly later...
	if version != "" {
		// Write channel into state
		k.ChannelKeeper.WriteOpenTryChannel(ctx, msg.PortId, channelID, msg.Channel.Ordering, msg.Channel.ConnectionHops, msg.Channel.Counterparty, version)
	}

	return &channeltypes.MsgChannelOpenTryResponse{
		Version: version,
	}, nil
```

This may also justify adding something like a `WriteOpenTryChannel` method to `ICS4Wrapper` to extend the async version feature to Middleware.

File: spec/core/ics-026-routing-module/README.md [hunk 3]
```diff
@@ -610,18 +652,29 @@ interface ChanCloseConfirm {
...
+    handler.chanCloseConfirm(
+      datagram.portIdentifier,
+      datagram.channelIdentifier,
+      datagram.proofInit,
+      datagram.proofHeight
+    )
+    // SYNCHRONOUS: the following calls happen syncrhonously with the call above
+    // ASYNCHRONOUS: the module callback will be called at a time later than the channel handler
+    // in this case, the channel identifier will be stored with a sentinel value so it is not taken
+    // by a new channel handshake and the capability is reserved for the application module.
```
Comment (crodriguezvega) [inline spec/core/ics-026-routing-module/README.md:665]: Is this actually needed?

File: spec/core/ics-026-routing-module/README.md [hunk 4]
```diff
@@ -447,6 +447,12 @@ function handleChanOpenInit(datagram: ChanOpenInit) {
     datagram.portIdentifier,
     datagram.counterpartyPortIdentifier
   )
+  // SYNCHRONOUS: the following calls happen synchronously with the call above
+  // ASYNCHRONOUS: the module callback will be called at a time later than the channel handler
+  // in this case, the channel identifier will be stored with a sentinel value so it is not taken
```
Comment (crodriguezvega) [inline spec/core/ics-026-routing-module/README.md:452]: With this you mean that a sentinel value is stored in the path `channelEnds/ports/{datagram.portIdentifier}/channels/{channelIdentifier}`, right? Maybe it's more explicit to mention it like this?

**Comment by AdityaSripal** [comment]
@michaelfig yes. We are trying to improve the interface for middleware to avoid the complicated wiring as it exists. I think this will be much cleaner in future designs, and the ICS4 wrapper interface itself should be unnecessary

**Pull request merged**